### ISSUE TITLE
UnicodeDecodeError when accessing my local Music folder

### ISSUE THREAD

**Issue by sticken88** [mopidy/mopidy#1502]
Hi everyone! I am a computer engineer and I'd like to contribute to Mopidy development! I installed it and started to play around with Gnome Music Player Client to understand how mopidy works.

When I tried to access my Music folder from GMPC I got the following error from mopidy:

``` python
ERROR    FileBackend backend caused an exception.
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/mopidy/core/library.py", line 19, in _backend_error_handling
    yield
  File "/usr/lib/python2.7/dist-packages/mopidy/core/library.py", line 112, in _browse
    result = backend.library.browse(uri).get()
  File "/usr/lib/python2.7/dist-packages/pykka/threading.py", line 52, in get
    compat.reraise(*self._data['exc_info'])
  File "/usr/lib/python2.7/dist-packages/pykka/compat.py", line 12, in reraise
    exec('raise tp, value, tb')
  File "/usr/lib/python2.7/dist-packages/pykka/actor.py", line 201, in _actor_loop
    response = self._handle_receive(message)
  File "/usr/lib/python2.7/dist-packages/pykka/actor.py", line 295, in _handle_receive
    return callee(*message['args'], **message['kwargs'])
  File "/usr/lib/python2.7/dist-packages/mopidy/file/library.py", line 67, in browse
    if not self._is_in_basedir(os.path.realpath(child_path)):
  File "/usr/lib/python2.7/dist-packages/mopidy/file/library.py", line 136, in _is_in_basedir
    for media_dir in self._media_dirs)
  File "/usr/lib/python2.7/dist-packages/mopidy/file/library.py", line 136, in <genexpr>
    for media_dir in self._media_dirs)
  File "/usr/lib/python2.7/dist-packages/mopidy/internal/path.py", line 213, in is_path_inside_base_dir
    common_prefix = os.path.commonprefix([real_base_path, real_path])
  File "/usr/lib/python2.7/genericpath.py", line 71, in commonprefix
    s1 = min(m)
UnicodeDecodeError: 'ascii' codec can't decode byte 0xe2 in position 31: ordinal not in range(128)
```

It seems that the server still works properly though, because the music streaming isn't blocked by this error. I'd like to fix it (or try to do so :) ) as a first step as contributor. What do you think?

I run Ubuntu 14.04 LTS and python 2.7.6 on my laptop.

**Comment by rmccawl** [comment]
Similar issue here, running OSX (10.11.4)

```
ERROR    'ascii' codec can't encode character u'\ufeff' in position 0: ordinal not in range(128)
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/site-packages/mopidy/__main__.py", line 134, in main
    return args.command.run(args, proxied_config)
  File "/usr/local/lib/python2.7/site-packages/mopidy/local/commands.py", line 150, in run
    track = tags.convert_tags_to_track(result.tags).replace(
  File "/usr/local/lib/python2.7/site-packages/mopidy/audio/tags.py", line 81, in convert_tags_to_track
    'musicbrainz-sortname')
  File "/usr/local/lib/python2.7/site-packages/mopidy/audio/tags.py", line 137, in _artists
    return [Artist(**attrs)]
  File "/usr/local/lib/python2.7/site-packages/mopidy/models/immutable.py", line 158, in __call__
    *args, **kwargs)
  File "/usr/local/lib/python2.7/site-packages/mopidy/models/immutable.py", line 34, in __init__
    self._set_field(key, value)
  File "/usr/local/lib/python2.7/site-packages/mopidy/models/immutable.py", line 186, in _set_field
    object.__setattr__(self, name, value)
  File "/usr/local/lib/python2.7/site-packages/mopidy/models/fields.py", line 50, in __set__
    value = self.validate(value)
  File "/usr/local/lib/python2.7/site-packages/mopidy/models/fields.py", line 98, in validate
    return compat.intern(str(super(Identifier, self).validate(value)))
UnicodeEncodeError: 'ascii' codec can't encode character u'\ufeff' in position 0: ordinal not in range(128)
Traceback (most recent call last):
  File "/usr/local/bin/mopidy", line 9, in <module>
    load_entry_point('Mopidy==2.0.0', 'console_scripts', 'mopidy')()
  File "/usr/local/lib/python2.7/site-packages/mopidy/__main__.py", line 134, in main
    return args.command.run(args, proxied_config)
  File "/usr/local/lib/python2.7/site-packages/mopidy/local/commands.py", line 150, in run
    track = tags.convert_tags_to_track(result.tags).replace(
  File "/usr/local/lib/python2.7/site-packages/mopidy/audio/tags.py", line 81, in convert_tags_to_track
    'musicbrainz-sortname')
  File "/usr/local/lib/python2.7/site-packages/mopidy/audio/tags.py", line 137, in _artists
    return [Artist(**attrs)]
  File "/usr/local/lib/python2.7/site-packages/mopidy/models/immutable.py", line 158, in __call__
    *args, **kwargs)
  File "/usr/local/lib/python2.7/site-packages/mopidy/models/immutable.py", line 34, in __init__
    self._set_field(key, value)
  File "/usr/local/lib/python2.7/site-packages/mopidy/models/immutable.py", line 186, in _set_field
    object.__setattr__(self, name, value)
  File "/usr/local/lib/python2.7/site-packages/mopidy/models/fields.py", line 50, in __set__
    value = self.validate(value)
  File "/usr/local/lib/python2.7/site-packages/mopidy/models/fields.py", line 98, in validate
    return compat.intern(str(super(Identifier, self).validate(value)))
UnicodeEncodeError: 'ascii' codec can't encode character u'\ufeff' in position 0: ordinal not in range(128)
```

**Comment by sticken88** [comment]
I definitely want to fix it..

**Comment by jodal** [comment]
The two errors are similar, but quite different: they happen in two different places, one while encoding and the other while decoding. Let's keep this bug for tracking @sticken88's crash.

@rmccawl Can you file a separate bug with your crash and what you did leading up to the crash? Can you reproduce it consistently?

**Comment by rmccawl** [comment]
Hi @jodal, Thanks for your reply. I'll create a new bug and the steps to reproduce on my end.

**Comment by jodal** [comment]
@rmccawl Great! Your issue looks like something interesting for a bugfix release as well :-)

**Comment by rmccawl** [comment]
Here's the separate bug report @jodal - https://github.com/mopidy/mopidy/issues/1508
If you need anymore info, let me know over on that issue.

Cheers.

**Comment by sticken88** [comment]
I think I got it. I submitted a pull request. @rmccawl maybe it can help you to fix issue #1508 :). Let me know!

**Comment by sticken88** [comment]
I forgot a quick explanation of the issue.

Basically the name of a folder had a dash encoded as "\xe2\x80\x93" so mopidy couldn't decode that and it raised the UnicodeDecodeError. My solution was to remove every non-ascii char using regexp.

**Comment by jjok** [comment]
@sticken88 Are these folders named after artists and albums? Presumably they could legitimately contain a lot of non-ascii characters (eg only Japanese characters)?

**Comment by sticken88** [comment]
@jjok you're right, I should consider that scenario as well. I think I have to provide another solution to fix the problem. Maybe we just need to specify the character encoding used by python?

I googled a bit and found this: http://stackoverflow.com/questions/2276200/changing-default-encoding-of-python

What do you guys think?

**Comment by jjok** [comment]
I would guess that you want to use UTF-8 encoding, although I don't know much about the problem.

**Comment by sticken88** [comment]
Yes that's my idea. All right I'll try to set the proper encoding and do some tests.

**Comment by kingosticks** [comment]
Paths in Mopidy are normally byte strings but it looks like the media_dirs might not be obeying this. Fixing that might be the real solution here.

**Comment by kingosticks** [comment]
Looking closer, `_get_media_dirs()` calls `path.expand_path()` which can return either unicode or bytes depending on your system configuration. On my system it expands `$XDG_MUSIC_DIR` to the unicode path `u'/home/nsteel/Music'`. Then later on we try to pass both a bytestring and unicode string `os.path.commonprefix()`, which it doesn't like, hence the exception.

Maybe life would be simpler if `path.expand_path()` did the conversion for us? @adamcik any reason why it doesn't? Don't we *always* want paths to be bytes?

**Comment by tkem** [comment]
AFAICS (after @kingosticks pointing me into the right direction), the root cause seems to be `mopidy.internal.xdg.get_user_dirs()` returning Unicode paths for `XDG_MUSIC_DIR` etc. when parsing a user's `$XDG_CONFIG_HOME/user-dirs.dirs`. So in theory at least, this should not happen when running Mopidy as a service.

The following commit seems relevant: 97c6b8812dc140e95eac01ba04a5bf6815bac90c

**Comment by sticken88** [comment]
So which solution should we adopt?

**Comment by kingosticks** [comment]
Either way is probably fine. Just out of principle, I'm voting for doing it once in `xdg.get_dirs()` rather than every time we call `path.expand_path()`.

**Comment by kingosticks** [comment]
Hang on, isn't this just a duplicate of https://github.com/mopidy/mopidy/pull/1345 which was actually fixed in develop ~2 months ago via https://github.com/mopidy/mopidy/pull/1493? That'll teach me to not update.

**Comment by sticken88** [comment]
So is this one already fixed?

**Comment by kingosticks** [comment]
I believe so. @jodal, am I missing something?

**Comment by adamcik** [comment]
Closing under the assumption that kingosticks is right about this just being #1345. Please reopen if you can reproduce in develop. Thanks.
